File: server-catalog/src/Catalog.Api/Middleware/RequestLoggingMiddleware.cs
using System.Diagnostics;

namespace Catalog.Api.Middleware;

/// <summary>
/// Outermost middleware — runs before everything else, including
/// <see cref="ExceptionHandlingMiddleware"/>, so every request gets exactly one structured
/// log line no matter how it ends. Generates a correlation id when the caller does not send
/// one, and echoes it back in the response header so a caller (or someone reading logs
/// later) can tie a client-side report to the exact server-side log line.
/// </summary>
public class RequestLoggingMiddleware
{
    private const string CorrelationIdHeader = "X-Correlation-Id";

    private readonly RequestDelegate _next;
    private readonly ILogger<RequestLoggingMiddleware> _logger;

    public RequestLoggingMiddleware(RequestDelegate next, ILogger<RequestLoggingMiddleware> logger)
    {
        _next = next;
        _logger = logger;
    }

    public async Task InvokeAsync(HttpContext context)
    {
        var correlationId = context.Request.Headers.TryGetValue(CorrelationIdHeader, out var existing) && !string.IsNullOrWhiteSpace(existing)
            ? existing.ToString()
            : Guid.NewGuid().ToString("n");

        context.Response.Headers[CorrelationIdHeader] = correlationId;

        var stopwatch = Stopwatch.StartNew();

        using (_logger.BeginScope(new Dictionary<string, object> { ["CorrelationId"] = correlationId }))
        {
            try
            {
                await _next(context);
            }
            finally
            {
                stopwatch.Stop();

                _logger.LogInformation(
                    "{Method} {Path} responded {StatusCode} in {ElapsedMs}ms",
                    context.Request.Method,
                    context.Request.Path,
                    context.Response.StatusCode,
                    stopwatch.ElapsedMilliseconds);
            }
        }
    }
}

File: server-orders/src/common/middleware/request-logging.middleware.ts
import { Injectable, NestMiddleware } from '@nestjs/common';
import { randomUUID } from 'crypto';
import { NextFunction, Request, Response } from 'express';
import { JsonLogger } from '../logging/json-logger';

const CORRELATION_ID_HEADER = 'x-correlation-id';

/**
 * Applied to every route in AppModule.configure() -- the same "outermost, wraps everything"
 * placement as server-catalog's RequestLoggingMiddleware. Generates or forwards a
 * correlation id, echoes it back in the response header, and logs one structured line per
 * request once it finishes (method, path, status, elapsed ms).
 */
@Injectable()
export class RequestLoggingMiddleware implements NestMiddleware {
  constructor(private readonly logger: JsonLogger) {}

  use(req: Request, res: Response, next: NextFunction): void {
    const correlationId = (req.headers[CORRELATION_ID_HEADER] as string) || randomUUID();
    res.setHeader('X-Correlation-Id', correlationId);

    const startedAt = process.hrtime.bigint();

    res.on('finish', () => {
      const elapsedMs = Number(process.hrtime.bigint() - startedAt) / 1_000_000;

      this.logger.log(
        {
          correlationId,
          method: req.method,
          path: req.originalUrl,
          statusCode: res.statusCode,
          elapsedMs: Math.round(elapsedMs * 100) / 100,
        },
        RequestLoggingMiddleware.name,
      );
    });

    next();
  }
}
